exercises/hamming/test/Tests.hs

```diff
@@ -102,3 +110,65 @@ cases = [ Case { description = "empty strands"
                , expected    = Nothing
                }
         ]
+
+hammingGen :: Int -> Gen Case
+hammingGen 0 = return
+  Case { description = "Empty strands have no distance"
+       , strand1 = ""
+       , strand2 = ""
+       , expected = Just 0
+       }
+
+hammingGen n = do
+  (a, b, count) <- nucleotidePairGen
+  case' <- hammingGen (n-1)
+  return . hammingDesc $
+    case' { description = ""
+          , strand1 = a ++ strand1 case'
+          , strand2 = b ++ strand2 case'
+          , expected = (+) <$> count <*> expected case'
+          }
+
+hammingDesc :: Case -> Case
+hammingDesc case'@Case{..} = case' { description = desc' }
+  where
+    desc' = "Strands of " ++ length' ++ " with " ++ dist' ++ " distance"
+    dist' = maybe "invalid" show expected
+    length' = case comparing length strand1 strand2 of
+      EQ -> "lengths " ++ show (length strand1)
+      _  -> "different lengths " ++ show (length strand1) ++ " and " ++
+            show (length strand2)
+
+nucleotidePairGen :: Gen (String, String, Maybe Integer)
+nucleotidePairGen = frequency
+  [ (20, zeroDistance)
+  , (4, oneDistance)
+  , (1, invalidDistance)
+  ]
+  where
+    zeroDistance = do
+      a <- elements "ACGT"
+      return ([a], [a], Just 0)
+
+    oneDistance = do
+      a <- elements "ACGT"
+      b <- elements ("ACGT" \\ [a])
+      return ([a], [b], Just 1)
+
+    invalidDistance = do
+      a <- elements "ACGT"
+      return ([a], [], Nothing)
+
+without :: [a] -> [(a, [a])]
+without [] = []
+without (x:xs) = (x, xs) : map (second (x:)) (without xs)
+
+hammingShrink :: Case -> [Case]
+hammingShrink case'@Case{..} = map hammingDesc $
+  zipWith shrink' (without strand1) (without strand2)
+  where
+    shrink' (a, strand1') (b, strand2')
+      | a == b = case' { strand1 = strand1', strand2 = strand2' }
```

Comment:
optional: I'm sure it'd be possible to write the `strand1 = strand1', strand2 = strand2'` once instead of twice as it it is now, but I'm not convinced that's necessary.